csrc/ir/internal_nodes.h

```diff
@@ -2327,4 +2327,77 @@ class LinearOp : public Expr {
   }
 };
 
+// SDPA node with same functionality as F.scaled_dot_product_attention
+// https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
```

Comment:
Thanks for raising this point. The current `SdpaOp` node will work only for inference.
For training: we would need to create two separate ops for forward and backward with different function signatures.

Another issue that @wujingyue pointed out is, that the current ATen evaluation does not look at whether the inputs are sharded which would need some modifications to the `evaluate` method. This is broader issue, since it applies to all other ops as well but particularly to `MatmulOp` and `LinearOp`, where we use the ATen fallback path primarily..

Reply:
Just want to point out that, for our ExprEvaluator, returning the intermediates (residuals?!) for backward isn't a big deal, but if we want to decompose and schedule it, we might want to be more careful about it.

Ideally speaking, nvfuser should be able to trim un-used outputs. So maybe we should switch, but which signature is the right one to target here? PyTorch itself has so many different variants and each has a different signature:
https://github.com/pytorch/pytorch/blob/a6b994ed5467d4df8320cbae51cba6a98ffb139c/aten/src/ATen/native/transformers/attention.cpp#L665-L706
https://github.com/pytorch/pytorch/blob/a6b994ed5467d4df8320cbae51cba6a98ffb139c/tools/autograd/derivatives.yaml#L2806-L2829

Reply:
I summarized the current plan we settled on here: https://github.com/NVIDIA/Fuser/issues/2278